lib/srv/db/postgres/users.go

```diff
@@ -329,28 +337,31 @@ func (e *Engine) DeleteUser(ctx context.Context, sessionCtx *common.Session) err
 	}
 	defer conn.Close(ctx)
 
-	e.Log.WithField("user", sessionCtx.DatabaseUser).Info("Deleting PostgreSQL user.")
+	logger := e.Log.WithField("user", sessionCtx.DatabaseUser)
+	logger.Info("Deleting PostgreSQL user.")
 
 	var state string
-	switch {
-	case sessionCtx.Database.IsRedshift():
-		err = e.deleteUserRedshift(ctx, sessionCtx, conn, &state)
-	default:
-		err = conn.QueryRow(ctx, deleteQuery, sessionCtx.DatabaseUser).Scan(&state)
-	}
+	err = withRetry(ctx, logger, func() error {
+		switch {
+		case sessionCtx.Database.IsRedshift():
```

Comment:
Assuming you have tested this and SQL state is properly returned for Redshift. 

@gabrielcorado I recall there were some issues capturing the SQL state from Redshift when trying to rethrow a state or something?

Reply:
yes I tested this, and the e2e test (without my deadlock fix) was retrying and passed despite deadlocks. I kept the deadlock fix because we shouldn't deadlock ourselves anyway.

The error comes back as a pgconn.PgError with the code set to the deadlock state

Reply:
as I understand it, the redshift delete procedure will rethrow the error 55006 (object_in_use) when it fails to delete a user and instead falls back to deactivating them in its exception handler, which we check for and then return our custom "user deactivated" state. This won't be retried because 1) it's not an error we consider "retryable" and 2) our code returns `nil` error for that case.

For deadlocks and serialization errors, a procedure's transaction is rolled back and a new transaction is started when it gets into the exception handler. The only redshift proc with an exception handler is the delete proc.
Once in the handler, if there are no further issues it will deactivate the user but it will still rethrow the deadlock exception to our pgx client. So we'll still see the deadlock or serialization error and retry user deletion proc in its entirety, allowing us to drop the user as intended.